### PR TITLE
ACNA-2839 | Added Update Template Endpoint

### PR DESCRIPTION
**Pull request by amulyakashyap09**
## Description

New endpoint for updating an existing template. This will be a new action in the service and it should be at the path PUT /templates/{templateId}

If the template does not exist, throw a 404 error.

## Endpoint behavior

1. Look to see if the passed template has credentials or apis
2. If present, just overwrite the passed request body as the new template
3. If not present, look for links.consoleProject and call the new /install-config endpoint for the project projectId. Add the returned api and credential config to the new template record
4. Add the new endpoint to openapi.yml definitions 

We'll also need to make sure the request body can handle the updatedBy field, that includes the username of the user who updated the template.

This endpoint should be protected by IMS auth and the t`emplate_registry.write scope`

## Proposal for update template endpoint

https://wiki.corp.adobe.com/display/ioarch/Proposal%3A+Expand+Template+Registry+to+support+Console+Templates#Proposal:ExpandTemplateRegistrytosupportConsoleTemplates-Updatetemplate

## Related Issue

[ACNA-2839](https://jira.corp.adobe.com/browse/ACNA-2839)

## How Has This Been Tested?

Locally on machine via unit-tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: template-registry-api.json [hunk 1]
```diff
@@ -578,9 +670,26 @@
...
+            "write:template_registry": "modify the template registry"
+          }
         }
       },
       "parameters": {
+        "TemplateId": {
+          "name": "templateId",
+          "in": "path",
+          "required": true,
+          "description": "The name of the template on npmjs.com",
```
Comment (arjuncooliitr) [inline template-registry-api.json:688]: Is templateId same as templateName? Else the description should say Id

File: template-registry-api.json [hunk 2]
```diff
@@ -1005,6 +1114,345 @@
...
           }
+        },
+        "updateTemplate": {
+          "type": "object",
+          "required": ["updatedBy"],
+          "additionalProperties": false,
+          "properties": {
+            "updatedBy": {
+              "type": "string",
+              "description": "The name of the user who created the template",
```
Comment (arjuncooliitr) [inline template-registry-api.json:1125]: Should be name of the user who updated

**Comment by shazron** [comment]
@amulyakashyap09 Tests are not passing.

File: template-registry-api.json [hunk 3]
```diff
@@ -489,6 +489,76 @@
...
+          },
+          "503": {
+            "$ref": "#/responses/ServiceUnavailable"
+          },
+          "504": {
+            "$ref": "#/responses/GatewayTimeout"
+          }
+        },
+        "x-openwhisk": {
+          "action": "templates-post.http",
```
Comment (MichaelGoberling) [inline template-registry-api.json:556]: needs to be templates-put.http

File: template-registry-api.json [hunk 4]
```diff
@@ -489,6 +489,76 @@
         }
       }
     },
+    "/templates/{templateId}": {
```
Comment (MichaelGoberling) [inline template-registry-api.json:492]: getting the following error when linting this spec: 

```
Paths should resolve unambiguously. Found two ambiguous paths: `/templates/{templateId}` and `/templates/{templateName}`.
```

File: actions/templates/put/index.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      return errorResponse(401, errorMessages, logger);
+    }
+
+    const isTemplateIdValid = PUT_PARAM_NAME in params && typeof params[PUT_PARAM_NAME] === 'string' && params[PUT_PARAM_NAME].length > 0;
+    if (!isTemplateIdValid) {
+      return errorResponse(400, [errorMessage(ERR_RC_MISSING_REQUIRED_PARAMETER, `The "${PUT_PARAM_NAME}" parameter is not set.`)], logger);
+    }
+
+    // extract the user Bearer token from the Authorization header
+    const accessToken = getBearerToken(params);
```
Comment (MichaelGoberling) [inline actions/templates/put/index.js:82]: access token validation is done at the runtime level now based on our swagger security definition, no need to validate again here

**Review by MichaelGoberling — CHANGES_REQUESTED** [review]

File: actions/templates/delete/index.js [hunk 6]
```diff
@@ -75,29 +104,17 @@ async function main (params) {
...
-
-    await removeTemplateByName(dbParams, fullTemplateName);
-    const response = {
-      statusCode: 200
-    };
-
     logger.info('"DELETE templates" executed successfully');
     return response;
   } catch (error) {
+    console.trace(error);
```
Comment (MichaelGoberling) [inline actions/templates/delete/index.js:117]: remove, `logger.error(error)` should capture what we need below

**Review by MichaelGoberling — APPROVED** [review]
LGTM besides one minor debug console log

**Pull request merged**